### DISCUSSION TITLE
Date Filter rule and readByQuery

### DISCUSSION THREAD

**Discussion by maxbritto** [directus/directus#15333] [Archive]
Hello,

I have an issue with filtering my items based on their date with the `ItemsService` and the `readByQuery` function.
I've created only 2 objects for my tests, one is in the past, the other in the future :
```
[
  {
    id: '8d83ec2d-b109-443e-ab0f-87ef5a1b2d7b',
    status: 'scheduled',
    date: '2023-08-01T12:00:00',
    type: 'publish'
  },
  {
    id: 'b2b16651-be55-4b5d-9bc0-a4c827c51f8b',
    status: 'scheduled',
    date: '2021-08-01T12:00:00',
    type: 'publish'
  }
]
```

If I use the REST API filter, it works great : `filter={ "date": { "_lte" : "$NOW"}}` and I only fetch the object with a date in the past

But if I use the `readByQuery` function with the same server/dataset I fetch the two objects listed above
```typescript
import { defineEndpoint } from '@directus/extensions-sdk';

export default defineEndpoint((router, { services }) => {
	const { ItemsService } = services;
	router.get('/', async (_req, res) => {
		const actionService = new ItemsService('action', { schema: _req.schema });
		const actions: Action[] = await actionService.readByQuery({
			filter: {
				date: {
					_lte: "$NOW"
				}
			}
		});
		console.log(actions);
	});
});
```

I've tried using `_gte` for debugging and it does inverse the one result on the REST API and return none of the objects with the `readByQuery` function. It's like it is considering they are both in the future when using the `readByQuery` function, even though one is in 2021.

What I am doing wrong here ?

Thanks for your help :)

PS : I am on Directus 9.16.1 running locally on a macos 12.5 and Node 18.8.0

**Comment by rijkvanzanten** [comment]
I'm pretty sure it's because the `readByQuery` expects the "parsed" query object rather than raw input you'd give to the REST API. The REST API will sanitize and parse the query before calling `readByQuery`, which in turn means that the `$NOW` value and other dynamic variables are resolved before it's going to the readByQuery call. 

In this particular case, you could just swap out the `$NOW` for a `new Date()`, as that's what the query sanitization would do anyways:

https://github.com/directus/directus/blob/99f38285fd58e7033a9fbb37f161e79f5d735b48/packages/shared/src/utils/parse-filter.ts#L74-L82

**Reply by maxbritto** [reply]
You were right I did not realized this at all. I actually studied the `parse-filter.ts` code yesterday before posting my question but I thought this code was used by the `readByQuery` function.
Thanks for your help, it works by replacing the `"$NOW"` by `new Date()`

I'd be willing to contribute to the documentation on this part. What would be the best way : add sections on this page https://docs.directus.io/extensions/creating-extensions.html or is there already a technical documentation for the classes of the extension-sdk and I could add content to it ?

**Reply by rijkvanzanten** [reply]
The dream is to use https://typedoc.org to auto-generate technical references from the actual source code, so we can put information like that in the JS Doc description of the method. That's a tremendous amount of work though, so in the meantime, the next best thing would be to add a note to the extensions docs you linked 🙂 

Alternatively, we could discuss an update to these internal services to have them do the query sanitization rather than the step before it. That way, it would work the way you expected it to in the first place 🙂
